### ISSUE TITLE
Complex branch: Results seem to 'accumulate' in memory

### ISSUE THREAD

**Issue by za419** [za419/Android-calculator#17]
That is to say, results seem to be done with old values every once in a while.

I will update this once I am able to isolate a sequence of commands that triggers this issue on a clean build.

**Comment by za419** [comment]
I still don't know what steps cause this to occur, but observation of an event where it occured prior to using pow() showed that it seems to manifest in a failure to update currentValue - This is diagnosed because pow() displays currentValue as a suggestion for an exponent, and it continued to display 45i after I entered i.

**Comment by za419** [comment]
I finally have a sequence of events that triggers this bug. On the current commit (fe73eef4d39c6916d3e852119ae881c4504ab60c), follow the following steps on a clean build:
1. Press i
2. Press "Raise To"
3. Press Calculate (using the default exponent, which should be i)
4. Press zero
5. Press i
6. Press Add
7. Enter 1
8. Press Calculate
9. Press "Raise To"
10. Press i
11. Press Calculate
12. Press i
13. Press Add
14. Enter 1
15. Press Calculate
16. The result will be 2.0+i, which is incorrect and seems to use the old value of 1.0+i for the base of the calculation (that is, it appears to perform 1.0+i+1 instead of i+1).

I will update this if I find that any simpler sequence triggers the bug.

**Comment by za419** [comment]
<h1>I FINALLY GET IT!</h1>

Okay, fair explanation time.

This problem was caused by one of the constants (especially ones used by parseComplex) being modified. The specific case I found caused the `Complex.I` constant to modify itself by an addition to 1. Therefore, when parseComplex parsed "i", it would return `Complex.I`, which now held the value of 1+i. 

To solve this, I need to finally bite the bullet and make it so that Complex operations do not modify the instance, instead returning a new one with the computation result. This has slightly worse performance, but it is worth it to avoid the bugs.

Additionally, in order to avoid further bugs, I have decided to make Complex instances immutable (except for epsilons). This is irritating, and a reversal of a decision I made when I first defined the class. However, since it is at least insanely difficult (if not impossible) to make an immutable instance of a mutable class, and I do want the Complex constants to be publicly available, this is my only sane option.

I am sad. But this will allow general code to handle Complex instances without worry of causing the same arcane bugs.
